File: ui-vue3/src/views/traffic/_shared/ruleVersion.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  listRuleVersionsAPI,
  type RuleVersion,
  type RuleVersionList,
  type TrafficRuleKind
} from '@/api/service/traffic'
import { HTTP_STATUS } from '@/base/http/constants'

export interface LatestRecordedState {
  versionNo?: number
  latestRecordedDeleted: boolean
}

export const latestRecordedStateFromItems = (items: RuleVersion[]): LatestRecordedState => {
  const latestRecorded = items.find((item) => item.isLatestRecorded)
  const head = items[0]
  return {
    versionNo: latestRecorded?.versionNo,
    latestRecordedDeleted: Boolean(head?.operation === 'DELETE')
  }
}

export const latestRecordedStateFromList = (list?: RuleVersionList): LatestRecordedState => {
  if (!list) {
    return { latestRecordedDeleted: false }
  }
  if (list.latestRecordedVersionNo !== undefined || list.latestRecordedDeleted !== undefined) {
    return {
      versionNo: list.latestRecordedVersionNo,
      latestRecordedDeleted: Boolean(list.latestRecordedDeleted)
    }
  }
  return latestRecordedStateFromItems(list.items || [])
}

export const versionDiffLabel = (prefix: string, versionNo?: number): string =>
  typeof versionNo === 'number' && versionNo > 0 ? `${prefix} v${versionNo}` : prefix

export const isLatestRecordedHistoryRequest = (
  requestSeq: number,
  latestSeq: number,
  disposed: boolean
) => {
  return !disposed && requestSeq === latestSeq
}

export const fetchLatestRecordedState = async (
  kind: TrafficRuleKind,
  ruleName: string
): Promise<LatestRecordedState> => {
  const res = await listRuleVersionsAPI(kind, ruleName)
  if (res.code === HTTP_STATUS.SUCCESS) {
    return latestRecordedStateFromList(res.data)
  }
  return { latestRecordedDeleted: false }
}

export const formatRuleSpec = (specJson?: string): string => {
  if (!specJson) {
    return ''
  }
  try {
    return JSON.stringify(JSON.parse(specJson), null, 2)
  } catch (e) {
    return specJson
  }
}

File: pkg/console/model/rule_version.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package model

import (
	"time"

	coremodel "github.com/apache/dubbo-admin/pkg/core/resource/model"
	"github.com/apache/dubbo-admin/pkg/core/versioning"
)

type RollbackRuleVersionReq struct {
	Reason string `json:"reason"`
}

type RuleVersionResp struct {
	RuleKind                coremodel.ResourceKind `json:"ruleKind"`
	Mesh                    string                 `json:"mesh"`
	ResourceKey             string                 `json:"resourceKey"`
	RuleName                string                 `json:"ruleName"`
	VersionNo               int64                  `json:"versionNo"`
	ContentHash             string                 `json:"contentHash"`
	SpecJSON                string                 `json:"specJson"`
	Source                  versioning.Source      `json:"source"`
	Operation               versioning.Operation   `json:"operation"`
	Author                  string                 `json:"author"`
	Reason                  string                 `json:"reason,omitempty"`
	RolledBackFromVersionNo *int64                 `json:"rolledBackFromVersionNo,omitempty"`
	CreatedAt               time.Time              `json:"createdAt"`
	RecordedAt              time.Time              `json:"recordedAt"`
	IsLatestRecorded        bool                   `json:"isLatestRecorded"`
}

type RuleVersionListResp struct {
	Items                   []RuleVersionResp `json:"items"`
	Total                   int64             `json:"total"`
	LatestRecordedVersionNo int64             `json:"latestRecordedVersionNo,omitempty"`
	LatestRecordedDeleted   bool              `json:"latestRecordedDeleted"`
}

type RuleVersionDiffResp struct {
	Left  RuleVersionDiffSideResp `json:"left"`
	Right RuleVersionDiffSideResp `json:"right"`
}

type RuleVersionDiffSideResp struct {
	VersionNo int64  `json:"versionNo"`
	SpecJSON  string `json:"specJson"`
}

type RollbackRuleVersionResp struct {
	RolledBackFromVersionNo int64  `json:"rolledBackFromVersionNo"`
	VersionNo               int64  `json:"versionNo"`
	Source                  string `json:"source"`
}
